rfc/0084-app-service-level-telemetry.md

```diff
@@ -0,0 +1,516 @@
+      },
+      "appTelemetryPath": "/_appTelemetry"
+    }
+
+## Network Interaction
+
+Once the SDK determined the endpoint it should establish HTTP connection and
+upgrade it to WebSocket.
+
+The server allows to close connection at any point of time.
+
+Once the server closes connection, the SDK should pick next available endpoint.
+Both initial and next endpoints must be randomized in case of more than one
+endpoint configured.
+
+The SDK must wait for backoff interval before connection attempt to avoid
+stressing the server. The SDK might use `app_telemetry_backoff` option to allow
+to change the backoff value.
+
+The SDK must either implement or enable WebSocket PING frame, and respond to
+server request with WebSocket PONG whenever the server requests it.
+
+The SDK must respond to any WebSocket PING frame with PONG frame whenever it is
+possible to avoid termination of the connection.
+
+The SDK might also send PING frames to the server and expect PONG frame. If the
+server does not respond to SDK-initiated PING request, the SDK should close
+connection, and reconnect after `app_telemetry_backoff` interval.
+
+## Wire Protocol
+
+The structure of the protocol is intentionally simple, and yet allows for
+future extension.
+
+The SDK leverages WebSocket framing, so does not need to have special fields
+for the payload size.
+
+First byte is reserved for Command codes in requests from the server, and for
+Status for SDK responses.
+
+This RFC defines the following types of the commands.
+
+| Name            | Value  | Description                                                                  |
+|-----------------|--------|------------------------------------------------------------------------------|
+| `GET_TELEMETRY` | `0x00` | Retrieve Application Telemetry from SDK in Prometheus Text Exposition Format |
+
+And the following statuses:
+
+| Name              | Value  | Description                                                |
+|-------------------|--------|------------------------------------------------------------|
+| `SUCCESS`         | `0x00` | No errors has been occurred                                 |
+| `UNKNOWN_COMMAND` | `0x01` | The SDK does not support operation requested by the server |
+
+
+### `GET_TELEMETRY`
+
+The request should not have payload.
+
+The response might have payload if the SDK has metrics to report.
+
+The SDK must reset local counters after successful scheduling response for the
+network delivery.
+
+The format of the response follows the [Text-based exposition format of the
+Prometheus](https://prometheus.io/docs/instrumenting/exposition_formats/#text-based-format),
+and includes metrics and histograms collected by the SDK since the previous
+report.
+
+#### Metrics
+
+    sdk_kv_r_timedout{agent="couchbase-net-sdk/2.4.5.0 (clr/4.0.30319.42000) (os/Microsoft Windows NT 10.0.16299.0)",node="node1",node_uuid="91442eb8202e0e16bbb59624d9ccdb0a",bucket="travel-sample"} 1 1695747260
```

Comment:
The RFC requires timestamps to be in millisecond-precision (which is consistent with the Prometheus spec), but the examples appear to be in seconds. It's not that important, but it might be worth updating them to avoid any confusion.

Reply:
updated